File: Advanced Programming Methods/toyLanguage -sleep/src/sample/model/OpenRFile.java
package sample.model;

import sample.utils.FileData;
import sample.utils.IdGenerator;
import sample.utils.MyIDictionary;
import sample.utils.exceptions.InterpretorException;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class OpenRFile implements IStmt {
    private String var_file_id;  //variable name
    private String fileName;

    public OpenRFile(String var_file_id, String fileName){
        this.fileName=fileName;
        this.var_file_id=var_file_id;
    }

    private void containsp(PrgState p){
        for (FileData f:p.getFileTable().getAllValues()) {
            if (f.getFileName().equals(fileName))
                throw new InterpretorException("The file is already open");
        }

    }

    @Override
    public PrgState execute(PrgState p){
        containsp(p);
        try{
            BufferedReader bf=new BufferedReader(new FileReader(fileName));
            FileData filed=new FileData(fileName, bf);
            int id= IdGenerator.generateId();
            p.getFileTable().add(id, filed);
            MyIDictionary<String, Integer> st=p.getSymTable();
            if(st.contains(var_file_id))
                st.update(var_file_id, id);
            else
                st.add(var_file_id, id);

        }catch (IOException e){
            throw new InterpretorException("The file cannot be opened. It does not exist or another error occured.");
        }
        return null;
    }

    @Override
    public String toString(){
        return var_file_id+" "+fileName;
    }
}

File: Advanced Programming Methods/toyLanguage/toyLanguage/model/OpenRFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using toyLanguage;
using System.IO;

namespace toyLanguage.model
{
    class OpenRFile:IStmt
    {
        private String var_file_id;
        private String fileName;
        public OpenRFile(String fid, String fn)
        {
            var_file_id = fid;
            fileName = fn;
        }

        public void Containsp(PrgState p)
        {
            try
            {
                IDictionary<int, FileData> d = p.FileTable.GetAll();
                    foreach (FileData fd in d.Values)
                    {
                        if (fd.FileName.Equals(fileName))
                        {
                            throw new Exception("The file is already open!");
                        }
                    }
               
            }
            catch(Exception e)
            {

            }
            
        }

        public PrgState Execute(PrgState p)
        {
            try
            {
                this.Containsp(p);
                StreamReader fs = new StreamReader("read.txt");
                FileData f = new FileData(fileName, fs);
                int id = IDGenerator.GenerateId();
                p.FileTable.Add(id, f);
                IMyDictionary<String, int> st = p.SymTable;
                if (st.ContainsKey(var_file_id))
                    st.Update(var_file_id, id);
                else
                    st.Add(var_file_id, id);
            }
           
            catch (NullReferenceException e)
            {
                Console.WriteLine(e.StackTrace);
            }
            return p; 
        }

        public override string ToString()
        {
            return "(OpenRFile (" + this.var_file_id + "," + this.fileName + "))";
        }
    }
}
